File: 15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    // let res_arr = {};
    // for (let i = 0; i < nums.length - 2; i++) {
    //     const num = nums[i];
    //     let res = twoSum(nums.slice(i + 1), -num);
    //     res.map(function(arr) {
    //         arr.push(num);
    //         arr.sort((a, b) => a - b);
    //         res_arr[arr.join(',')] = arr;
    //     });
    // }

    // let res = [];
    // for (const key in res_arr) {
    //     res.push(res_arr[key]);
    // }

    // return res;

    // function twoSum(nums, value) {
    //     let res = [];
    //     for (let i = 0; i < nums.length - 1; i++) {
    //         let num = nums[i];
    //         let res_arr = nums.slice(i + 1);
    //         let j = res_arr.indexOf(value - num);
    //         if (~j) {
    //             res.push([num, res_arr[j]]);
    //         }
    //     }
    //     return res;
    // }

    //top rated
    var rtn = [];
    if (nums.length < 3) {
        return rtn;
    }
    nums = nums.sort(function(a, b) {
        return a - b;
    });
    for (var i = 0; i < nums.length - 2; i++) {
        if (nums[i] > 0) {
            return rtn;
        }
        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }
        for (var j = i + 1, k = nums.length - 1; j < k; ) {
            if (nums[i] + nums[j] + nums[k] === 0) {
                rtn.push([nums[i], nums[j], nums[k]]);
                j++;
                k--;
                while (j < k && nums[j] == nums[j - 1]) {
                    j++;
                }
                while (j < k && nums[k] == nums[k + 1]) {
                    k--;
                }
            } else if (nums[i] + nums[j] + nums[k] > 0) {
                k--;
            } else {
                j++;
            }
        }
    }
    return rtn;
};

// threeSum([-1, 0, 1, 2, -1, -4]);

File: 15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from typing import List
# @lc code=start


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        rtn = []
        # 先排序
        nums.sort()
        # 枚举 a
        for i in range(0, n):
            # 需要与前一次不同
            if i > 0 and nums[i] == nums[i-1]:
                continue
            # c 指针指向数组的最右端
            k = n -1
            target = -nums[i]
            for j in range(i + 1, n):
                # 需要与上一次枚举的数不同
                if j > i + 1 and nums[j] == nums[j -1 ]:
                    continue
                while j < k and nums[j] + nums[k] > target:
                    k -= 1
                # 如果指针重合，随着 b 后续的增加
                # 就不会有满足 a+b+c=0 并且 b<c 的 c 了，可以退出循环
                if j == k:
                    break;
                if nums[j] + nums[k] == target:
                    rtn.append([nums[i], nums[j], nums[k]])
        return rtn

# @lc code=end


sol = Solution()
res = sol.threeSum([-1,0,1,2,-1,-4])
print(res)
